src/generator.rs

```diff
@@ -568,8 +583,108 @@ mod tests {
+            "2a00:1450:4007:810::2004",
+            42254,
+            443,
+        );
+        let mut flow_information_2 = FlowInformation::new();
+        flow_information_2.sni = Some("mtalk.google.com".to_string());
+        let mut flag_list = BTreeSet::new();
+        flag_list.insert(Flag::ACK);
+        flow_information_2.backward_packet_list.push(Packet {
+            length: 218,
+            timestamp: Duration::new(1595324883, 969259000),
+            // assumed clone, we provide a copy
+            flag_list: flag_list.clone(),
+            network_protocol: 34525,
+            network_header_length: None,
+            position: 182,
+        });
+        flow_information_2.backward_packet_list.push(Packet {
+            length: 882,
+            timestamp: Duration::new(1595324884, 248056000),
+            // assumed clone, we provide a copy
+            flag_list: flag_list.clone(),
+            network_protocol: 34525,
+            network_header_length: None,
+            position: 196,
+        });
+        flow_information_2.forward_packet_list.push(Packet {
+            length: 558,
+            timestamp: Duration::new(1595324883, 910142000),
+            // assumed clone, we provide a copy
+            flag_list: flag_list.clone(),
+            network_protocol: 34525,
+            network_header_length: None,
+            position: 178,
+        });
+        flow_information_2.forward_packet_list.push(Packet {
+            length: 64,
+            timestamp: Duration::new(1595324884, 36342000),
+            // assumed clone, we provide a copy
+            flag_list: flag_list.clone(),
+            network_protocol: 34525,
+            network_header_length: Some(5),
+            position: 189,
+        });
+        flag_list.insert(Flag::CWR);
+        flag_list.insert(Flag::ECE);
+        flow_information_2.forward_packet_list.push(Packet {
+            length: 234,
+            timestamp: Duration::new(1595324884, 158344000),
+            // last flag list, we move
+            flag_list,
+            network_protocol: 34525,
+            network_header_length: None,
+            position: 194,
+        });
+        generator.add(flow_id_2, flow_information_2);
+
+        assert_eq!(generator.len(), 2);
+
+        let expected_flow_id_1 = FlowId::new(132, "10.216.31.192", "10.15.194.132", 0, 0);
+        let expected_flow_id_2 = FlowId::new(
+            6,
+            "2a01:cb06:a02d:8571:4706:7df1:bd62:5169",
+            "2a00:1450:4007:810::2004",
+            42254,
+            443,
+        );
+        let mut n = 0;
+        for (i, (flow_id, _flow_information)) in generator.into_iter().enumerate() {
+            n = i;
+            let result = panic::catch_unwind(|| {
+                assert_eq!(flow_id, expected_flow_id_1);
+            });
+            if result.is_err() {
+                assert_eq!(flow_id, expected_flow_id_2);
+            }
+        }
+        assert_eq!(n + 1, 2);
```

Comment:
A few func prog to simplify the code (no counter):

```
assert_eq!(
            generator.into_iter().fold(0, |n, (flow_id, _flow_information)| {
                let result = panic::catch_unwind(|| {
                    assert_eq!(flow_id, expected_flow_id_1);
                });
                if result.is_err() {
                    assert_eq!(flow_id, expected_flow_id_2);
                };
                n + 1
            }),
            2
        );
```

fold is your friend ;).

Reply:
👍